### ISSUE TITLE
Add command to rebuild plugins.php

### ISSUE THREAD

**Issue by sjelfull** [craftcms/cms#2793]
### Description

Related to #2419 - is it possible to add a command that reads the plugins the same way the custom plugin installer does, and thus recreate the plugins.php file on demand?

**Comment by dsmrt** [comment]
I'm running into an issue with the plugins.php and I think we need a service method so we can pass a plugin to it. 

_Context_
I am working on a plugin with unit and functional tests using Codeception. I have a base install of a craft db dump and would like to test plugin install.

_Example_
```php
if (! \Craft::$app->plugins->isPluginInstalled($plugin->getHandle())) {
    \Craft::$app->plugins->installPlugin($plugin->getHandle());
}
``` 

Since I'm running `composer install` from within the plugin repo it's not being added to the `vendor/craftcms/plugins.php`. I guess what ever composer event the craft plugin installer is hooking into isn't being triggered. 

_Error_
```bash
1) MyRecordTest: Simple record
 Test  tests/unit/MyRecordTest.php:testSimpleRecord

  [craft\errors\InvalidPluginException] No plugin exists with the handle "awesome-plugin".

#1  /var/www/html/plugin/vendor/craftcms/cms/src/services/Plugins.php:852
#2  /var/www/html/plugin/vendor/craftcms/cms/src/services/Plugins.php:481
#3  /var/www/html/plugin/tests/_support/Step/Unit/Common/AwesomePlugin.php:22
#4  /var/www/html/plugin/tests/unit/MyRecordTest.php:61
```

I can see the use of a command/cli for this but it'd be great to have a service like so:
```php
\Craft::$app->plugins->register($plugin)
```

If there is a current work around, any help would be welcome.

**Comment by lemiwinkz** [comment]
> I'm running into an issue with the plugins.php and I think we need a service method so we can pass a plugin to it.
> 
> _Context_
> I am working on a plugin with unit and functional tests using Codeception. I have a base install of a craft db dump and would like to test plugin install.
> 
> _Example_
> 
> ```
> if (! \Craft::$app->plugins->isPluginInstalled($plugin->getHandle())) {
>     \Craft::$app->plugins->installPlugin($plugin->getHandle());
> }
> ```
> 
> Since I'm running `composer install` from within the plugin repo it's not being added to the `vendor/craftcms/plugins.php`. I guess what ever composer event the craft plugin installer is hooking into isn't being triggered.
> 
> _Error_
> 
> ```shell
> 1) MyRecordTest: Simple record
>  Test  tests/unit/MyRecordTest.php:testSimpleRecord
> 
>   [craft\errors\InvalidPluginException] No plugin exists with the handle "awesome-plugin".
> 
> #1  /var/www/html/plugin/vendor/craftcms/cms/src/services/Plugins.php:852
> #2  /var/www/html/plugin/vendor/craftcms/cms/src/services/Plugins.php:481
> #3  /var/www/html/plugin/tests/_support/Step/Unit/Common/AwesomePlugin.php:22
> #4  /var/www/html/plugin/tests/unit/MyRecordTest.php:61
> ```
> 
> I can see the use of a command/cli for this but it'd be great to have a service like so:
> 
> ```
> \Craft::$app->plugins->register($plugin)
> ```
> 
> If there is a current work around, any help would be welcome.

+1 On this one. There is a formal Craft CMS [testing framework ](https://github.com/craftcms/cms/tree/feature/tests) on the way (E.T.A [DotAll 2019](https://dotall.com/2019)). I am helping build this. The framework will provide what you seek as native functionality - registration of a plugin will go through codeception.yaml. See the method that does this in the framework [here](https://github.com/craftcms/cms/blob/feature/tests/src/test/Craft.php#L267) (currently, its a proof of concept and will be refactored)- this method registers plugins if they are at the root of the projects directory (So not in vendor directory).

It's currently not ready for use (seriously - things are still being changed on the daily) but a service like this will be really useful and part of refactoring the above method - I will give it a whack this weekend to build a way that cleanly allows you to add plugins like you describe. Currently, plugins.php is modified by [this]( https://github.com/craftcms/plugin-installer/blob/master/src/Installer.php#L96).

@angrybrad Any ideas on this?

**Comment by brandonkelly** [comment]
Rather than running `composer install` from within the plugin repo, you should be linking a normal Craft project to your plugin’s repo by adding a [path repository](https://docs.craftcms.com/v3/extend/plugin-guide.html#path-repository) to its `composer.json` file and then running `composer install` from there.

**Comment by lemiwinkz** [comment]
> Rather than running `composer install` from within the plugin repo, you should be linking a normal Craft project to your plugin’s repo by adding a [path repository](https://docs.craftcms.com/v3/extend/plugin-guide.html#path-repository) to its `composer.json` file and then running `composer install` from there.

For @dsmrt 's point I agree, use the path repo's. Its a cleaner solution for testing your code whilst doing active development.

In the case of a Craft CMS plugin testing framework, my argument for supporting plugins at root is a CI system. 

Wouldnt it be quite tedious to have to set up an entire Craft project in a CI environment just to test a plugin - including things such as editing the composer.json file as the Craft docs describe? 

Seeing as CI systems (I.E Travis CI) allow you to inject [env variables](https://docs.travis-ci.com/user/languages/php/#testing-against-multiple-versions-of-dependencies) plugin devs should set up their tests based on what testing environment, CI or manual, currently is used/desired. For I.E Codeception they would do that in the [settings > bootstrap](https://codeception.com/docs/reference/Configuration) option in the `codeception.yml` file. Example of this script can be found [here](https://github.com/lemiwinkz/commerce/blob/develop/tests/_bootstrap.php#L7). Reason for this is because the plugin framework (as it stands now) needs a `CRAFT_VENDOR_PATH` constant which is different when plugins are the root of a project `vs` when the plugin is a composer package. 

Is there a better alternative for this that I am missing? If so we might as well drop support for plugins at root in `feature/tests`? 

@brandonkelly

**Comment by dsmrt** [comment]
You are spot on @lemiwinkz . Our codeception bootstraps are configure and working great (@nateiler built a lot of what I'm using). But pulling in an entire craft project to run tests on the plugin is a pain with CI. We are currently using TravisCI. 

I've played around with loading craft as a project into one of our [docker images](https://hub.docker.com/r/flipbox/craftcms) but it's kind of a hack and a lot of work. Also, whatever I run locally, I want the process to be as similar to the CI environment. Travis does a bunch of heavy lifting for me that I'd like to avoid. I'd rather not manage a bunch of docker images just for testing.

Here a glimpse at my docker entrypoint to show you the lengths I go to to automate including a plugin from the plugin repo to craft project within the docker image. There's a good amount of editing needed on the project composer file get things working.

```bash
#!/bin/bash

if [ "" != "${PLUGIN_REQUIRE}" ] && [ "" != "${PLUGIN_VERSION}" ]; then
    cat composer.json | jq '.require |= . + { "'$PLUGIN_REQUIRE'":"'$PLUGIN_VERSION'" }' | \
        jq '. |= . + {"minimum-stability": "dev"}' | \
        jq '. |= . + {"prefer-stable" : true}' | \
        jq '. |= . + {"repositories": [{"type": "composer","url": "https://asset-packagist.org"},{"type": "path","url": "'$PLUGIN_PATH'"}]}' > composer.temp

    cp composer.temp composer.json && rm -f composer.temp
    composer update
    cat composer.json | jq .
    chmod 777 -R /var/www/html/storage /var/www/html/config /var/www/html/web
fi

#pass it on to the docker-endpoint
exec docker-entrypoint "$@"
```

Plugin testing needs to be easier and the plugins.php is my major blocker right now. Everything else is working great.

**Comment by angrybrad** [comment]
Yeah, I think this will greatly help testing Craft/plugins in a CI environment.

Additionally, there are semi-frequent support tickets around this error:

> [craft\errors\InvalidPluginException] No plugin exists with the handle "awesome-plugin".

Where (for whatever reason) the plugins.php file gets out of sync with the database and Composer. This command would be a relatively simple way to get things back in sync.

**Comment by lemiwinkz** [comment]
@dsmrt `craftcms/plugin-installer` is the package that sets up `plugins.php`. [This PR ](https://github.com/craftcms/plugin-installer/pull/2) was just merged which adds support for plugins at root. Just ensure you have `1.5.3` of `craftcms/plugin-installer` and `composer install` will ensure your plugin is added to the `plugins.php` file.

**Comment by gtettelaar** [comment]
@dsmrt Just an F.Y.I seeing as you were working on testing a Craft plugin: 

As of Craft [3.2 BETA-1](https://github.com/craftcms/cms/blob/3.2/CHANGELOG-v3.md#320-beta1---2019-06-04), there is now a formalized testing Framework for Craft. Based on Codeception with support for sites, modules and plugins in various testing styles.

Formal announcement [here](https://twitter.com/CraftCMS/status/1135969539970048000)

Big +1 for @boboldehampsink for his help on this.

**Comment by jamierumbelow** [comment]
I'm not quite sure how – I had a problem with an upgrade to Composer v2 – but I've just managed to get my plugins.php out of sync. I deleted vendor and reinstalled my dependencies and it's broken.

A way to regenerate this opaque file from the DB would be a life-saver; as it stands I'm pulling the file from the production server and replacing it manually. Craft shouldn't be storing state in a directory one can consider ephemeral, like vendor.
